# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Marking.cpp
# Commit: 27d9b7db5e1c
# Full Hash: 27d9b7db5e1c50af9456023d63e2d3db72b48df2
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-10-28 04:43:55
# Regressor Bug: 1797755
# File Overlap Count: 1
# Description:
#   Bug 1797755 - Part 5: Use a single initial mark stack size regardless of whether incremental GC is enabled r=sfink
#   
#   Currently we initialize the mark stack to a different size depending on whether
#   or not incremental GC is enable. However, after the first GC we always shrink
#   it to the initial size when it is disabled.
# ==============================================================================

diff -r f207beed6252 -r 27d9b7db5e1c js/src/gc/Marking.cpp
--- a/js/src/gc/Marking.cpp	Thu Oct 27 18:23:03 2022 +0000
+++ b/js/src/gc/Marking.cpp	Thu Oct 27 18:23:03 2022 +0000
@@ -1567,18 +1567,13 @@
   MOZ_ASSERT(iteratorCount_ == 0);
 }
 
-bool MarkStack::init(bool incrementalGCEnabled) {
+bool MarkStack::init() {
   MOZ_ASSERT(isEmpty());
-  return setStackCapacity(incrementalGCEnabled);
+  return resetStackCapacity();
 }
 
-bool MarkStack::setStackCapacity(bool incrementalGCEnabled) {
-  size_t capacity;
-  if (incrementalGCEnabled) {
-    capacity = INCREMENTAL_MARK_STACK_BASE_CAPACITY;
-  } else {
-    capacity = NON_INCREMENTAL_MARK_STACK_BASE_CAPACITY;
-  }
+bool MarkStack::resetStackCapacity() {
+  size_t capacity = MARK_STACK_BASE_CAPACITY;
 
 #ifdef JS_GC_ZEAL
   capacity = std::min(capacity, maxCapacity_.ref());
@@ -1601,6 +1596,14 @@
 }
 #endif
 
+void MarkStack::clear() {
+  // Fall back to the smaller initial capacity so we don't hold on to excess
+  // memory between GCs.
+  stack().clearAndFree();
+  std::ignore = resetStackCapacity();
+  topIndex_ = 0;
+}
+
 inline MarkStack::TaggedPtr* MarkStack::topPtr() { return &stack()[topIndex_]; }
 
 inline bool MarkStack::pushTaggedPtr(Tag tag, Cell* ptr) {
@@ -1744,10 +1747,7 @@
 {
 }
 
-bool GCMarker::init() {
-  bool incrementalGCEnabled = runtime()->gc.isIncrementalGCEnabled();
-  return stack.init(incrementalGCEnabled);
-}
+bool GCMarker::init() { return stack.init(); }
 
 bool GCMarker::isDrained() { return isMarkStackEmpty() && !delayedMarkingList; }
 